[compiler-rt][cmake] Add option to control shared library builds

kateinoigakukun opened the pull request:
This patch introduces a new option `COMPILER_RT_BUILD_SHARED_LIBS` to control whether shared libraries should be built for runtime libraries in compiler-rt. By default, this option is set to ON, allowing shared libraries to be built. If set to OFF, shared libraries will not be built, which can be useful for projects that do not require them or when building for environments where shared libraries are still not yet matured, such as WebAssembly/WASI.

dschuff commented on compiler-rt/lib/asan/CMakeLists.txt:
```diff
@@ -1,5 +1,12 @@
 # Build for the AddressSanitizer runtime support library.
 
+set(COMPILER_RT_ASAN_BUILD_SHARED_LIBS_default ON)
+if (CMAKE_SYSTEM_NAME STREQUAL "WASI")
```
@sbc100 I assume we've never tried the ASan CMake build for emscripten since we have the standalone build scripts?
Anyway this might be a case where we want the same logic for WASI and Emscripten.

kateinoigakukun commented on compiler-rt/lib/asan/CMakeLists.txt:
```diff
@@ -1,5 +1,12 @@
 # Build for the AddressSanitizer runtime support library.
 
+set(COMPILER_RT_ASAN_BUILD_SHARED_LIBS_default ON)
```
Having a higher-level control sounds good to me. Since the compiler-rt libraries are built outside of the standard `add_library` CMake convention, it's not possible to disable all shared library build by just setting `BUILD_SHARED_LIBS` unfortunately.

Also it looks like building a shared library is currently broken due to https://github.com/llvm/llvm-project/issues/103592. Do you have a plan to merge https://github.com/llvm/llvm-project/issues/103592?

```
wasm-ld: error: lib/sanitizer_common/CMakeFiles/RTSanitizerCommon.wasm32.dir/sanitizer_wasi.cpp.obj: undefined symbol: __heap_base
wasm-ld: error: lib/sanitizer_common/CMakeFiles/RTSanitizerCommon.wasm32.dir/sanitizer_wasi.cpp.obj: undefined symbol: __heap_end
wasm-ld: error: lib/sanitizer_common/CMakeFiles/RTSanitizerCommon.wasm32.dir/sanitizer_wasi.cpp.obj: undefined symbol: __heap_base
wasm-ld: error: lib/sanitizer_common/CMakeFiles/RTSanitizerCommon.wasm32.dir/sanitizer_wasi.cpp.obj: undefined symbol: __heap_end
wasm-ld: error: lib/sanitizer_common/CMakeFiles/RTSanitizerCommon.wasm32.dir/sanitizer_wasi.cpp.obj: undefined symbol: __heap_base
wasm-ld: error: lib/sanitizer_common/CMakeFiles/RTSanitizerCommon.wasm32.dir/sanitizer_wasi.cpp.obj: undefined symbol: __heap_end
wasm-ld: error: lib/sanitizer_common/CMakeFiles/RTSanitizerCommonLibc.wasm32.dir/sanitizer_common_libcdep.cpp.obj: undefined symbol: __sanitizer::PlatformPrepareForSandboxing(void*)
wasm-ld: error: lib/sanitizer_common/CMakeFiles/RTSanitizerCommonSymbolizer.wasm32.dir/sanitizer_stackdepot.cpp.obj: undefined symbol: __sanitizer::internal_join_thread(void*)
wasm-ld: error: lib/asan/CMakeFiles/RTAsan_dynamic.wasm32.dir/asan_allocator.cpp.obj: undefined symbol: __global_base
wasm-ld: error: lib/asan/CMakeFiles/RTAsan_dynamic.wasm32.dir/asan_descriptions.cpp.obj: undefined symbol: __global_base
wasm-ld: error: lib/asan/CMakeFiles/RTAsan_dynamic.wasm32.dir/asan_descriptions.cpp.obj: undefined symbol: __global_base
wasm-ld: error: lib/asan/CMakeFiles/RTAsan_dynamic.wasm32.dir/asan_errors.cpp.obj: undefined symbol: __global_base
wasm-ld: error: lib/asan/CMakeFiles/RTAsan_dynamic.wasm32.dir/asan_errors.cpp.obj: undefined symbol: __global_base
wasm-ld: error: lib/asan/CMakeFiles/RTAsan_dynamic.wasm32.dir/asan_errors.cpp.obj: undefined symbol: __global_base
wasm-ld: error: lib/asan/CMakeFiles/RTAsan_dynamic.wasm32.dir/asan_errors.cpp.obj: undefined symbol: __global_base
wasm-ld: error: lib/asan/CMakeFiles/RTAsan_dynamic.wasm32.dir/asan_fake_stack.cpp.obj: undefined symbol: __global_base
wasm-ld: error: lib/asan/CMakeFiles/RTAsan_dynamic.wasm32.dir/asan_fake_stack.cpp.obj: undefined symbol: __global_base
wasm-ld: error: lib/asan/CMakeFiles/RTAsan_dynamic.wasm32.dir/asan_fake_stack.cpp.obj: undefined symbol: __global_base
wasm-ld: error: lib/asan/CMakeFiles/RTAsan_dynamic.wasm32.dir/asan_fake_stack.cpp.obj: undefined symbol: __global_base
wasm-ld: error: lib/asan/CMakeFiles/RTAsan_dynamic.wasm32.dir/asan_fake_stack.cpp.obj: undefined symbol: __global_base
wasm-ld: error: too many errors emitted, stopping now (use -error-limit=0 to see all errors)
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
```

kateinoigakukun replied:
I need to check how it will work with the `./runtimes` unified build but sounds good to me.

sbc100 reviewed (commented):
Could we have `add_compiler_rt_runtime` honor `BUILD_SHARED_LIBS` internally perhaps?

dschuff commented:
> Could we have `add_compiler_rt_runtime` honor `BUILD_SHARED_LIBS` internally perhaps?

I don't think this is a good idea. BUILD_SHARED_LIBS is about how the compiler is built, and that should be separate from how the runtime libraries are built. Also BUILD_SHARED_LIBS is a (LLVM) developer feature and it's well known that it shouldn't be used for release builds, but that's not necessarily true of the library build.

kateinoigakukun commented:
I agree with @dschuff's point about `BUILD_SHARED_LIBS`. Considering the unified build of the compiler and runtime libraries, using the same option to control sharedness for both the compiler and runtimes could result in building runtime libraries always only for static (as most distributions use BUILD_SHARED_LIBS=FALSE).

How about making `add_compiler_rt_runtime` to respect `COMPILER_RT_BUILD_SHARED_LIBS` instead?

sbc100 commented:
> > Could we have `add_compiler_rt_runtime` honor `BUILD_SHARED_LIBS` internally perhaps?
> 
> I don't think this is a good idea. BUILD_SHARED_LIBS is about how the compiler is built, and that should be separate from how the runtime libraries are built. Also BUILD_SHARED_LIBS is a (LLVM) developer feature and it's well known that it shouldn't be used for release builds, but that's not necessarily true of the library build.

Actually you are completely right.   compiler-rt is the *output* of the compiler, not the compiler itself.  So in this case its perfectly reasonable for a compiler running on wasi use use shared libraries versions of compiler-rt.  Since clang it a cross compiler there could be any number of compiler-rt  shared libraries built, even if the version of wasi you are using doesn't support shared libraries the compiler target (e.g. linux) could/does.

kateinoigakukun committed:
[compiler-rt][cmake] Add option to control shared library builds

This patch introduces a new option `COMPILER_RT_BUILD_SHARED_LIBS` to
control whether shared libraries should be built for runtime libraries
in compiler-rt. By default, this option is set to ON, allowing shared
libraries to be built. If set to OFF, shared libraries will not be
built, which can be useful for projects that do not require them
or when building for environments where shared libraries are still
not yet matured, such as WebAssembly/WASI.

sbc100 commented on compiler-rt/CMakeLists.txt:
```diff
@@ -91,6 +91,13 @@ mark_as_advanced(COMPILER_RT_SCUDO_STANDALONE_BUILD_SHARED)
+set(COMPILER_RT_BUILD_SHARED_LIBS_DEFAULT ON)
+if(CMAKE_SYSTEM_NAME STREQUAL "WASI")
+  set(COMPILER_RT_BUILD_SHARED_LIBS_DEFAULT OFF)
+endif()
```
I guess if you want to avoid building shared library version of compiler-rt for wasi you would need to check the specific target that compiler-rt is being built for (not CMAKE_SYSTEM_NAME which is where the build is running I think).

However, maybe just better to wait for a fix for the downstream issues with building wasm shared libraries.

kateinoigakukun replied:
The `CMAKE_SYSTEM_NAME` variable here points to the target platform where the runtime library runs on, not the platform where compiler runs. The LLVM CMake infrastructure uses `ExternalProject_Add` in the unified build of compiler + runtime libs, and it passes through most of CMake args **except for `CMAKE_SYSTEM_NAME`** and a few other exceptions. So CMake scripts in compiler-rt can assume `CMAKE_SYSTEM_NAME` is the target platform, not host.

kateinoigakukun replied:
Hmm, according to the CI failure, it seems like compiler-rt for the host platform is configured at the same CMake unit of the host compiler, so `BUILD_SHARED_LIBS` is shared for compiler and runtime libs for the host platform.
I think it's better to have a separate CMake variable to make things simpler.

petrhosek replied:
Using `BUILD_SHARED_LIBS` is the correct way. `LLVM_ENABLE_PROJECTS=compiler-rt` is not the right way to build compiler-rt for the host platform and we'll be [removing support for building compiler-rt that way after LLVM 21 branches](https://github.com/llvm/llvm-project/blob/83fbd86909111510f973d8e1c4214022368ef810/llvm/CMakeLists.txt#L188-L193).

kateinoigakukun replied:
Thanks! Then we should update the CI to use `LLVM_ENABLE_RUNTIMES` anyway

dschuff commented:
I feel like there must already be some kind of option to control whether shared or static libs are built. Maybe this is a question for a runtimes expert like @petrhosek

kateinoigakukun commented:
@petrhosek Would you mind taking a look at this if you have a moment? Thanks in advance :pray:
